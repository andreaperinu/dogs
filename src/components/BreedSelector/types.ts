import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface IBreedSelectorProps {
	maybeSelectedBreed: O.Option<IBreed>,
	maybeBreeds: O.Option<IBreed[]>,
	onChange: (maybeBreed: O.Option<IBreed>) => void
}
