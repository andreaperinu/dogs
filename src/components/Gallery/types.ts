import * as O from 'fp-ts/lib/Option';

export interface IGalleryProps {
	maybeImages: O.Option<string[]>
}
