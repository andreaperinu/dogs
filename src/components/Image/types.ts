import * as O from 'fp-ts/lib/Option'

export interface IImageProps {
	maybeImage: O.Option<string>
	fixedSize?: boolean
}
