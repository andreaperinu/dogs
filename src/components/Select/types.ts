import * as O from 'fp-ts/lib/Option';
import { BaseComponent } from '../../types';

export interface ISelectProps extends BaseComponent {
	maybeValue: O.Option<ISelectItem>
	maybeOptions: O.Option<ISelectItem[]>
	maybeText?: O.Option<string>
	disabled?: boolean
	onChange: (item: O.Option<ISelectItem>) => void
}

export interface ISelectItem {
	id: string
	value: string
}

export interface ISelectStyledProps {
	showOptions: boolean
	disabled: boolean
}
