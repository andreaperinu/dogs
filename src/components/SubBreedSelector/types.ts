import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface ISubBreedSelectorProps {
	disabled?: boolean
	maybeSelectedSubBreed: O.Option<string>,
	maybeBreed: O.Option<IBreed>,
	onChange: (maybeSubBreed: O.Option<string>) => void
}

export interface ISubBreedSelectorStyledProps {
	disabled: boolean
}
