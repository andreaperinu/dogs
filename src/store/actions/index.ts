export { fetchBreeds, fetchRandomImage, resetRandomImage, fetchImages, resetImages } from './actions'
