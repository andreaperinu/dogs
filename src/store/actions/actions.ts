import {
	FETCH_BREEDS, FETCH_IMAGES, FETCH_RANDOM_IMAGE, RESET_IMAGES, RESET_RANDOM_IMAGE
} from '@/store/actions/rawActions';
import { Action } from '@/store/types';

export type BreedPayload = { breed: string }

export const fetchBreeds: Action = () => ({ type: FETCH_BREEDS })
export const fetchRandomImage: Action<BreedPayload> = (payload) => (
	{ type: FETCH_RANDOM_IMAGE, payload }
)

export const resetRandomImage: Action = () => ({ type: RESET_RANDOM_IMAGE })
export const resetImages: Action = () => ({ type: RESET_IMAGES })

export const fetchImages: Action<BreedPayload> = (payload) => ({ type: FETCH_IMAGES, payload })
