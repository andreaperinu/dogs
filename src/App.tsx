import './App.css';

import React, { lazy, Suspense, useEffect, useMemo } from 'react';
import { Redirect, Route, Switch, useLocation } from 'react-router-dom';

import { Navigation } from '@/components/Navigation/Navigation';

import { AppStyled } from './App.styled';
import { Loader } from './components/Loader/Loader';
import { fetchBreeds, resetImages, resetRandomImage } from './store/actions';
import { useStore } from './store/store';

const BreedRandomPage = lazy(async () => import('@/pages/BreedRandomPage/BreedRandomPage'))
const SubBreedRandomPage = lazy(async () => import('@/pages/SubBreedRandomPage/SubBreedRandomPage'))
const BreedListPage = lazy(async () => import('@/pages/BreedListPage/BreedListPage'))
const SubBreedListPage = lazy(async () => import('@/pages/SubBreedListPage/SubBreedListPage'))

export const App = () => {
  const dispatch = useStore(false)[1]

  const { pathname } = useLocation()

  useEffect(() => {
    dispatch(fetchBreeds())
  }, [dispatch])

  useEffect(() => {
    dispatch(resetRandomImage())
    dispatch(resetImages())
  }, [pathname])

  return (
    <AppStyled>
      <header>
        <h1>Dog dashboard</h1>
      </header>

      <main>
        <Navigation />

        <Suspense fallback={<Loader />}>
          <Switch>
            <Route exact path="/list" component={BreedListPage} />
            <Route exact path="/sub/list" component={SubBreedListPage} />
            <Route exact path="/sub/random" component={SubBreedRandomPage} />
            <Route exact path="/" component={BreedRandomPage} />
            <Redirect to="/" />
          </Switch>
        </Suspense>
      </main>

      <footer>
        Made with ♥️ using Typescript, React, fp-ts, styled-components
      </footer>
    </AppStyled>
  );
}
