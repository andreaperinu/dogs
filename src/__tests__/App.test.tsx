import React from 'react';
import { act } from 'react-dom/test-utils';
import reactRouter from 'react-router';

import { FETCH_BREEDS, RESET_IMAGES, RESET_RANDOM_IMAGE } from '@/store/actions/rawActions';
import { setup } from '@/tests/utils';
import * as O from 'fp-ts/lib/Option'
import { App as component } from '../App';
import { IStoreState } from '@/store/types';

const dispatch = jest.fn()

beforeEach(() => {
	reactRouter.useLocation = jest.fn().mockReturnValue({ pathname: '' })
})

const initialState: Partial<IStoreState> = {
	maybeBreeds: O.none,
	maybeRandomImage: O.none,
}

describe('<App />', () => {
	it('should match snapshot', () => {
		const wrapper = setup({ component, props: {} })
		expect(wrapper).toMatchSnapshot()
	})

	it('should fetch breeds', () => {
		jest.spyOn(React, 'useEffect').mockImplementation(f => f())

		setup({
			component, props: {}, initialState, dispatch, shouldMount: true, shouldHaveRouter: true
		})

		expect(dispatch).toHaveBeenCalledWith({ type: FETCH_BREEDS })
	})

	it('should dispatch reset random image', () => {
		const wrapper = setup({
			component, props: {}, initialState, dispatch, shouldMount: true, shouldHaveRouter: true
		})

		void act(() => { wrapper.unmount() })

		expect(dispatch).toHaveBeenCalledWith({ type: RESET_RANDOM_IMAGE })
	})

	it('should dispatch reset images', () => {
		const wrapper = setup({
			component, props: {}, initialState, dispatch, shouldMount: true, shouldHaveRouter: true
		})

		void act(() => { wrapper.unmount() })

		expect(dispatch).toHaveBeenCalledWith({ type: RESET_IMAGES })
	})
})
