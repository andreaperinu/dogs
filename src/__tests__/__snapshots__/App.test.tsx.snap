// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match snapshot 1`] = `
<styled.div>
  <header>
    <h1>
      Dog dashboard
    </h1>
  </header>
  <main>
    <Navigation />
    <Suspense
      fallback={<styled.div />}
    >
      <Switch>
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.lazy),
              "_init": [Function],
              "_payload": Object {
                "_result": [Function],
                "_status": -1,
              },
            }
          }
          exact={true}
          path="/list"
        />
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.lazy),
              "_init": [Function],
              "_payload": Object {
                "_result": [Function],
                "_status": -1,
              },
            }
          }
          exact={true}
          path="/sub/list"
        />
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.lazy),
              "_init": [Function],
              "_payload": Object {
                "_result": [Function],
                "_status": -1,
              },
            }
          }
          exact={true}
          path="/sub/random"
        />
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.lazy),
              "_init": [Function],
              "_payload": Object {
                "_result": [Function],
                "_status": -1,
              },
            }
          }
          exact={true}
          path="/"
        />
        <Redirect
          to="/"
        />
      </Switch>
    </Suspense>
  </main>
  <footer>
    Made with ♥️ using Typescript, React, fp-ts, styled-components
  </footer>
</styled.div>
`;
